src/dev-app/focus-trap/focus-trap-demo.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {FocusTrap, FocusTrapFactory} from '@angular/cdk/a11y';
+import {
+  AfterViewInit,
+  Component,
+  ElementRef,
+  ViewChild,
+  ViewEncapsulation} from '@angular/core';
+import {MatDialog} from '@angular/material/dialog';
+
+@Component({
+  selector: 'shadow-dom-demo',
+  template: '<ng-content></ng-content>',
+  host:     {'class': 'demo-focus-trap-shadow-root'},
+  encapsulation: ViewEncapsulation.ShadowDom
+})
+export class FocusTrapShadowDOMDemo {}
+
+@Component({
+  selector: 'focus-trap-demo',
+  templateUrl: 'focus-trap-demo.html',
+  styleUrls: ['focus-trap-demo.css'],
+})
+export class FocusTrapDemo implements AfterViewInit {
+
+  basicFocusTrap: FocusTrap;
+  @ViewChild('basicDemoRegion', {static: false}) private readonly _basicDemoRegion!: ElementRef;
+
+  nestedOuterFocusTrap: FocusTrap;
+  @ViewChild('nestedOuterDemoRegion', {static: false})
+  private readonly _nestedOuterDemoRegion!: ElementRef;
+  nestedInnerFocusTrap: FocusTrap;
+  @ViewChild('nestedInnerDemoRegion', {static: false})
+  private readonly _nestedInnerDemoRegion!: ElementRef;
+
+  tabIndexFocusTrap: FocusTrap;
+  @ViewChild('tabIndexDemoRegion', {static: false})
+  private readonly _tabIndexDemoRegion!: ElementRef;
+
+  shadowDOMFocusTrap: FocusTrap;
+  @ViewChild('shadowDOMDemoRegion', {static: false})
+  private readonly _shadowDOMDemoRegion!: ElementRef;
+
+  iframeFocusTrap: FocusTrap;
+  @ViewChild('iframeDemoRegion', {static: false})
+  private readonly _iframeDemoRegion!: ElementRef;
+
+  dynamicFocusTrap: FocusTrap;
+  @ViewChild('dynamicDemoRegion', {static: false})
+  private readonly _dynamicDemoRegion!: ElementRef;
+  @ViewChild('newElements', {static: false}) private readonly _newElements!: ElementRef;
+
+  constructor(
+    public dialog: MatDialog,
+    private _focusTrapFactory: FocusTrapFactory) {}
+
+  ngAfterViewInit() {
+    setTimeout(() => {
```

Comment:
Add a comment that explains why this `setTimeout` is necessary?

Reply:
It was to avoid ExpressionChangedAfterViewHasBeenChecked errors because I was applying a class like [class.demo-focus-trap-enabled]="basicFocusTrap && basicFocusTrap.enabled", and that was switching from undefined before the FocusTrap was initialized to false when I created and disabled it.

I changed it to [class.demo-focus-trap-enabled]="(basicFocusTrap && basicFocusTrap.enabled) || false" to stop the errors without using setTimeout.